src/core/dbus-service.c

```diff
@@ -91,6 +94,70 @@ static int property_get_exit_status_set(
         return sd_bus_message_close_container(reply);
 }
 
+int bus_service_method_bind_mount(sd_bus_message *message, void *userdata, sd_bus_error *error) {
+        int read_only, make_file_or_directory;
+        const char *dest, *src, *propagate_directory;
+        Unit *u = userdata;
+        ExecContext *c;
+        int r;
+
+        assert(message);
+        assert(u);
+
+        if (!MANAGER_IS_SYSTEM(u->manager))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED, "Adding bind mounts at runtime is only supported for system managers.");
+
+        r = mac_selinux_unit_access_check(u, message, "start", error);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_read(message, "ssbb", &src, &dest, &read_only, &make_file_or_directory);
+        if (r < 0)
+                return r;
+
+        if (!path_is_absolute(src) || !path_is_normalized(src))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Source path must be absolute and normalized.");
+
+        if (isempty(dest))
+                dest = src;
+        else if (!path_is_absolute(dest) || !path_is_normalized(dest))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Destination path must be absolute and normalized.");
+
+        r = bus_verify_manage_units_async_full(
+                        u,
+                        "bind-mount",
+                        CAP_SYS_ADMIN,
+                        N_("Authentication is required to bind mount on '$(unit)'."),
+                        true,
+                        message,
+                        error);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
+
+        if (u->type != UNIT_SERVICE)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Unit is not of type .service");
+
+        /* If it would be dropped at startup time, return an error */
+        c = unit_get_exec_context(u);
+        if (c && path_startswith_strv(dest, c->inaccessible_paths))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "%s is not accessible to this unit", dest);
+
+        if (SERVICE(u)->state != SERVICE_RUNNING)
+                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Unit is not running");
```

Comment:
hmm, i think we should have another check here, that does a high-level check if namespaces are used (i mean, i am aware that bind_mount_in_namespace() does a low-level check, but I'd catch this earlier, given that daemons can do their own namespacing if they like and we should stay away then if we can detect that)

So, one option would be to generalize  exec_needs_mount_namespace() so that it works with runtime/param object set.

Or, to make this simpler, maybe we should insist on PrivateMounts= being set and then just check for that? i.e. use PrivateMounts= as the explicit opt-in for allowing the bind operation to work?  I mean, that's not too much to ask tat people opt-into this stuff explicitly by setting this variable?

Reply:
ok, made exec_needs_mount_namespace generally available (in src/core) by making exec_params optional - I figure that is fine to do, as it's only used for the prefix? Then we have the checks for the other dozens of directives that result in a namespaced execution